File: binarysearchtree.py
class Node(object):
    def __init__(self, element=None, parent=None):
        self.element = element
        self.parent = parent
        self.left = None
        self.right = None

    def get_element(self):
        return self.element

    def __str__(self):
        return str(self.element)

    def get_minimum(self, root):
        if root is None:
            return None
        if root.left is None:
            return root
        return self.get_minimum(root.left)

    def get_next_node_climb(self, start):
        if start.parent is None:
            return None
        parent = start.parent
        if (parent.left is not None) and (parent.left is start):
            return parent
        return self.get_next_node_climb(parent)

    def get_next_node(self):
        if self.right is not None:
            return self.get_minimum(self.right)
        return self.get_next_node_climb(self)

    def __iter__(self):
        self.next_node = self.get_minimum(self)
        return self

    def __next__(self):
        return_node = self.next_node
        if return_node is None:
            raise StopIteration
        self.next_node = self.next_node.get_next_node()
        return return_node.element


class BinarySearchTree(object):
    def __init__(self, name, root):
        self.name = name
        self.root = None
        self.size = 0

    def add_aux(self, r, elem, parent):
        if r is None:
            self.size += 1
            return Node(elem, parent)

        # if r.element is None:
        # self.size += 1
        # return r

        if elem < r.get_element():
            r.left = self.add_aux(r.left, elem, r)

        elif elem > r.get_element():
            r.right = self.add_aux(r.right, elem, r)

        return r

    def add(self, element):
        self.root = self.add_aux(self.root, element, None)

    def add_all(self, *values):
        for x in values:
            self.add(x)
        if self.root is None:
            self.root = Node()

    def print_node(self, node):
        if node is None:
            return ""
        if (node.left is None) and (node.right is None):
            result = "" + str(node.element)
            return result
        elif (node.left is None) and (node.right is not None):
            result = "" + str(node.element) + " R:(" + self.print_node(node.right) + ")"
            return result
        elif (node.left is not None) and (node.right is None):
            result = "" + str(node.element) + " L:(" + self.print_node(node.left) + ")"
            return result
        result = "" + str(node.element) + " L:(" + self.print_node(node.left) + ")" + " R:(" + self.print_node(
            node.right) + ")"
        return result

    def __str__(self):
        result = "[" + self.name + "] " + self.print_node(self.root)
        return result

    def get_minimum_2(self, root):
        if root is None:
            return None
        if root.left is None:
            return root
        return self.get_minimum_2(root.left)

    def __iter__(self):
        self.next_node = self.get_minimum_2(self.root)
        return self

    def __next__(self):
        return_node = self.next_node
        if return_node is None:
            raise StopIteration
        self.next_node = self.next_node.get_next_node()
        return return_node.element


class Merger(object):
    @staticmethod
    def merge(tree1, tree2):
        counter1 = 0
        counter2 = 0
        it1 = iter(tree1)
        it2 = iter(tree2)
        new1 = False
        new2 = False
        result_list = []
        while len(result_list) < tree1.size + tree2.size:
            if (new1 is False) and (counter1 < tree1.size):
                val1 = next(it1)
                counter1 += 1
                new1 = True

            if (new2 is False) and (counter2 < tree2.size):
                val2 = next(it2)
                counter2 += 1
                new2 = True

            if (new1 is False) and (new2 is True):
                result_list.append(val2)
                new2 = False
            elif (new1 is True) and (new2 is False):
                result_list.append(val1)
                new1 = False
            elif val1 < val2:
                result_list.append(val1)
                new1 = False
            elif val1 > val2:
                result_list.append(val2)
                new2 = False
            else:
                result_list.append(val1)
                new1 = False
        return result_list

File: BinarySearchTree.java
import java.util.*;

public class BinarySearchTree<T extends Comparable<T>> implements Iterable<T>{
    private String name;
    private BinarySearchTreeNode<T> rootNode;
    private int treeSize = 0;

    public BinarySearchTree(String name){
        this.name = name;
        this.rootNode = null;
    }

    public void addAll(List<T> elements){
        for (int i = 0; i < elements.size(); i++){
            add(elements.get(i));
        }
    }

    private void add(T element){
        rootNode = addAux(rootNode, element, null);
    }

    private BinarySearchTreeNode<T> addAux(BinarySearchTreeNode<T> r, T element, BinarySearchTreeNode<T> parent){
        if (r == null){
            treeSize++;
            return new BinarySearchTreeNode<>(element, parent);
        }
        if (element.compareTo(r.element) < 0) {r.left = addAux(r.left, element , r);}
        else if (element.compareTo(r.element) > 0) {r.right = addAux(r.right, element, r);}

        return r;
    }
    @Override
    public Iterator<T> iterator() {
        Iterator<T> it = new Iterator<T>() {
            BinarySearchTreeNode<T> nextNode = getMinimum2(rootNode);
            @Override
            public boolean hasNext() {
                return nextNode != null;
            }

            @Override
            public T next() {
                BinarySearchTreeNode<T> returnNode = nextNode;
                nextNode = nextNode.getNextNode();
                return returnNode.element;
            }
        };
        return it;
    }

    public BinarySearchTreeNode<T> getMinimum2(BinarySearchTreeNode<T> root){
        if (root == null){
            return null;
        }
        if (root.left == null){
            return root;
        }
        return getMinimum2(root.left);
    }

    public String toString(){
        String result = "[" + name + "] " + printNode(rootNode);
        return result;
    }

    private String printNode(BinarySearchTreeNode<T> node){
        if(node == null){return "";}

        if(node.left == null && node.right == null){
            String result = "" + node.element;
            return result;
        }

        else if(node.left == null && node.right != null){
            String result = "" + node.element + " R:(" + printNode(node.right) + ")";
            return result;
        }

        else if(node.left != null && node.right == null){
            String result = "" + node.element + " L:(" + printNode(node.left) + ")";
            return result;
        }

        String result = "" + node.element + " L:(" + printNode(node.left) + ")" + " R:(" + printNode(node.right) + ")";
        return result;

    }

    public class BinarySearchTreeNode<T extends Comparable<T>> {
        T element;
        BinarySearchTreeNode<T> parent;
        BinarySearchTreeNode<T> left;
        BinarySearchTreeNode<T> right;

        public BinarySearchTreeNode(T element){
            this.element = element;
        }

        public BinarySearchTreeNode(T element, BinarySearchTreeNode<T> parent){
            this.element = element;
            this.parent = parent;
        }

        public BinarySearchTreeNode<T> getMinimum(BinarySearchTreeNode<T> root){
            if (root == null){
                return null;
            }
            if (root.left == null){
                return root;
            }
            return getMinimum(root.left);
        }

        public BinarySearchTreeNode<T> getNextNodeClimb(BinarySearchTreeNode<T> start){
            if (start.parent == null){
                return null;
            }
            BinarySearchTreeNode<T> parent = start.parent;
            if (parent.left != null && parent.left.equals(start)){
                return parent;
            }
            return getNextNodeClimb(parent);
        }


        public BinarySearchTreeNode<T> getNextNode(){
            if (right != null){
                return getMinimum(right);
            }
            return getNextNodeClimb(this);
        }


    }

    public static class MergedList<T extends Comparable<T>>{
        private ArrayList<T> mergedList;
        boolean newVal1 = false;
        boolean newVal2 = false;
        private T val1;
        private T val2;
        private Iterator<T> it1;
        private Iterator<T> it2;
        int doneMergedListSize;
        boolean done = false;



        public MergedList(BinarySearchTree<T> t1, BinarySearchTree<T> t2){
            mergedList = new ArrayList<>();
            it1 = t1.iterator();
            it2 = t2.iterator();
            this.doneMergedListSize = t1.treeSize + t2.treeSize;
        }

        public synchronized void insert(){
            if (val1 != null && val2 == null){
                mergedList.add(val1);
                newVal1 = false;
            }

            else if (val1 == null && val2 != null){
                mergedList.add(val2);
                newVal2 = false;
            }

            else if (val1.compareTo(val2) < 0){
                mergedList.add(val1);
                newVal1 = false;
            }
            else if (val1.compareTo(val2) > 0){
                mergedList.add(val2);
                newVal2 = false;
            }
            else {
                mergedList.add(val1);
                newVal1 = false;
            }

            if (mergedList.size() >= doneMergedListSize){
                done = true;
            }
            //notifyAll();
        }

        public synchronized void give1(){
            if (mergedList.size() >= doneMergedListSize){
                done = true;
            }
            while (newVal1 && !done) try{wait();} catch(InterruptedException ignore){}

            newVal1 = true;
            if (it1.hasNext()){
                val1 = it1.next();
            }
            else {
                val1 = null;
            }

            if ((newVal1 && newVal2)  && !(mergedList.size() >= doneMergedListSize)){
                insert();
            }
            notifyAll();

        }

        public synchronized void give2(){
            if (mergedList.size() >= doneMergedListSize){
                done = true;
            }
            while (newVal2 && !done) try{wait();} catch(InterruptedException ignore){}

            newVal2 = true;
            if (it2.hasNext()){
                val2 = it2.next();
            }
            else {
                val2 = null;
            }

            if ((newVal1 && newVal2) && !(mergedList.size() >= doneMergedListSize)){
                insert();
            }
            notifyAll();
        }

        public List<T> getMergedList(){
            return mergedList;
        }


    }

    public static <T extends Comparable<T>> List<T> merge(BinarySearchTree<T> t1, BinarySearchTree<T> t2){
        MergedList<T> resultList = new MergedList<>(t1,t2);
        Thread thread1 = new Thread(new Runnable(){
            @Override
            public void run() {
                while(!resultList.done){
                    resultList.give1();
                }
            }
        }
        );

        Thread thread2 = new Thread(new Runnable(){
            @Override
            public void run() {
                while(!resultList.done){
                    resultList.give2();
                }
            }
        }
        );
        thread1.start();
        thread2.start();

        try {
            thread1.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        try {
            thread2.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        return resultList.getMergedList();
    }


    public static void main(String... args) {
        // each tree has a name, provided to its constructor
        BinarySearchTree<Integer> t1 = new BinarySearchTree<>("Oak");
        // adds the elements to t1 in the order 5, 3, 0, and then 9
        t1.addAll(Arrays.asList(5, 3, 0, 9));
        BinarySearchTree<Integer> t2 = new BinarySearchTree<>("Maple");
        // adds the elements to t2 in the order 9, 5, and then 10

        t2.addAll(Arrays.asList(9, 5, 10));
        //t2.addAll(Arrays.asList(9, 5, 10));


        System.out.println(t1); // see the expected output for exact format

        t1.forEach(System.out::println); // iteration in increasing order

        System.out.println(t2); // see the expected output for exact format

        t2.forEach(System.out::println); // iteration in increasing order

        BinarySearchTree<String> t3 = new BinarySearchTree<>("Cornucopia");
        t3.addAll(Arrays.asList("coconut", "apple", "banana", "plum", "durian",
                "no durians on this tree!", "tamarind"));
        System.out.println(t3); // see the expected output for exact format

        t3.forEach(System.out::println); // iteration in increasing order

        List<Integer> merged = merge(t1, t2);
        merged.forEach(System.out::println);
        System.out.println(merged);
    }
}
